### ISSUE TITLE
Switching from a high `num_ctx` to a model with a low `num_ctx` causes cuda out of memory errors

### ISSUE THREAD

**Issue by jmorganca** [ollama/ollama#1871]
When switching from a large context window to a small one (a high `num_ctx` to a low `num_ctx`), Ollama will error due to out of memory. It seems that it will incorrectly try to re-allocate the same amount of memory as before (vs a new, smaller amount).

**Comment by iplayfast** [comment]
I wonder if that's what's causing https://github.com/jmorganca/ollama/issues/1691

**Comment by dhiltgen** [comment]
Unfortunately, from the CUDA perspective, the memory allocations aren't leaked (they seem to be tracked, and cleaned up on exit) so it's some higher-level logic error in llama.cpp.  

(multiple iterations, and I see the `nvidia-smi` memory usage slowly climbing...)
```
compute-sanitizer --tool memcheck  --leak-check full ./ollama-linux-amd64 serve
========= COMPUTE-SANITIZER
time=2024-02-27T22:41:19.218Z level=INFO source=images.go:710 msg="total blobs: 20"
time=2024-02-27T22:41:19.219Z level=INFO source=images.go:717 msg="total unused blobs removed: 0"
time=2024-02-27T22:41:19.221Z level=INFO source=routes.go:1019 msg="Listening on 127.0.0.1:11434 (version 0.1.27-11-g076237b)"
time=2024-02-27T22:41:19.221Z level=INFO source=payload_common.go:107 msg="Extracting dynamic libraries..."
time=2024-02-27T22:41:21.958Z level=INFO source=payload_common.go:146 msg="Dynamic LLM libraries [rocm_v6 cpu cpu_avx cpu_avx2 cuda_v11 rocm_v5]"
time=2024-02-27T22:41:21.958Z level=INFO source=gpu.go:94 msg="Detecting GPU type"
time=2024-02-27T22:41:21.958Z level=INFO source=gpu.go:265 msg="Searching for GPU management library libnvidia-ml.so"
time=2024-02-27T22:41:21.959Z level=INFO source=gpu.go:311 msg="Discovered GPU libraries: [/usr/lib/x86_64-linux-gnu/libnvidia-ml.so.545.23.08]"
time=2024-02-27T22:41:21.987Z level=INFO source=gpu.go:99 msg="Nvidia GPU detected"
...

{"function":"update_slots","level":"INFO","line":1680,"msg":"slot released","n_cache_tokens":820,"n_ctx":1024,"n_past":819,"n_system_tokens":0,"slot_id":0,"task_id":0,"tid":"140381368845888","timestamp":1709076897,"truncated":true}
[GIN] 2024/02/27 - 23:34:57 | 200 |         7m25s |       127.0.0.1 | POST     "/api/generate"
========= LEAK SUMMARY: 0 bytes leaked in 0 allocations
========= ERROR SUMMARY: 0 errors
```

**Comment by dhiltgen** [comment]
This should be resolved by #3218
